Pulsar add-on documentation

Keksoj opened the pull request:
As described in #660, we need to provide good documentation to use the pulsar addon (currently in beta version)

Todos:

- [x] brief presentation
- [x] add-on version
- [x] add-on policies
  - [x] storage
  - [x] retention
  - [x] offload
- [x] how to create an add-on
- [x] how to connect to it
  - [x] biscuit examples (rust, java)
  - [x] pulsarctl usage
  - [x] biscuit attenuation (not a priority) with biscuit-cli
- [x] plan & pricing
- [x] add-on section in the environment reference page

KannarFr commented:
Nice job @Keksoj , I will add parts to this PR too.

Keksoj commented:
@KannarFr yeah, the `pulsarctl` parts would be usefull, you know those better than I do.

KannarFr commented:
@Keksoj @urcadox @gaelreyrol can you review my commits?

@Keksoj  can you provide a rust & java quick usage in a single file? And what do you mean by "how to create an addon"', you want to do a gif on the console, cli commands? I think it's a good idea, feel free.

KannarFr commented:
@gaelreyrol can you put contextual alert blocks when you think it's mandatory as a new reader and potentially not aware of how pulsar works?

gaelreyrol commented:
> @Galimede can we embed the pulsar pricing component in this doc page?

The component first need to be released according to the [CHANGELOG](https://github.com/CleverCloud/clever-components/blob/master/CHANGELOG.md) and then a dedicated [shortcode](https://github.com/CleverCloud/hugo-cc-doc-theme/blob/master/layouts/shortcodes/pricingAddon.html) in the theme.

gaelreyrol commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,259 @@
+
+Tenants and namespaces allow for grouping and subgrouping of topics.
+
+What a pulsar add-on basically is, is a tenant/namespace pair that allows you to create and use as many topics as you wantt following this pattern:
```
I find difficult to read that one:
`What a pulsar add-on basically is, is`

gaelreyrol replied:
```suggestion
What a pulsar add-on basically is, is a tenant/namespace pair that allows you to create and use as many topics as you want following this pattern:
```

gaelreyrol commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,259 @@
+
+## Authorization
+
+Pulsar addon uses (Biscuit for Pulsar)[https://github.com/CleverCloud/biscuit-pulsar/commits/master] implementation which is directly pluggable to the Pulsar authentication and authorization system. Each addon exposes its own biscuit token.
```
```suggestion
Pulsar add-on uses (Biscuit for Pulsar)[https://github.com/CleverCloud/biscuit-pulsar] implementation which is directly pluggable to the Pulsar authentication and authorization system. Each add-on exposes its own biscuit token.
```

gaelreyrol commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,259 @@
+- cold storage
+- queue
+- hot storage
+- biscuit
```
```suggestion
- biscuit
- rabbitmq
- kafka
```

gaelreyrol replied:
As discussed with @cnivolle, we need to have `rabbitmq` & `kafka` keywords around Pulsar ;)

gaelreyrol commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,259 @@
+
+They are several modes of subscription. A consumer may subscribe exclusively, or share the subscription with other consumers.
+
+- Exclusive
```
What is this list ?

KannarFr replied:
subscription type list, I added clear title.

gaelreyrol commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,259 @@
+
+### Usage
+
+We advise you to use (`pulsarctl`)[https://github.com/streamnative/pulsarctl] provided by StreamNative. Here is an example to list topics in your addon (in your namespace):
```
```suggestion
We advise you to use (`pulsarctl`)[https://github.com/streamnative/pulsarctl] provided by StreamNative. Here is an example to list topics in your add-on (in your namespace):
```

gaelreyrol commented on deploy/addon/pulsar.md:
````diff
@@ -0,0 +1,259 @@
+We advise you to use (`pulsarctl`)[https://github.com/streamnative/pulsarctl] provided by StreamNative. Here is an example to list topics in your addon (in your namespace):
+
+```bash
+pulsarctl --admin-service-url https://<PULSAR_REST_URL> --auth-params <BISCUIT> --auth-plugin org.apache.pulsar.client.impl.auth.AuthenticationToken topics list <tenant>/<namespace>
````
You should use the add-on environment variables.

gaelreyrol commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,259 @@
+
+As Biscuit is a token, you can use `AuthenticationToken("<BISCUIT>")` provided by clients librairies to authenticate to our clusters without any tweak.
+
+* C++ client
```
I think should simply link this page: https://pulsar.apache.org/docs/en/client-libraries/

gaelreyrol commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,259 @@
+
+### Operations
+
+The pulsar addon given biscuit enable you to run several operations on namespace, its policies and the related topics.
```
```suggestion
The pulsar add-on given biscuit enable you to run several operations on namespace, its policies and the related topics.
```

gaelreyrol commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,259 @@
+
+### Attenuation
+
+The pulsar addon given biscuit can be attenuated, here is an attenuation example using (biscuit-cli)[https://github.com/biscuit-auth/biscuit-cli] from the given biscuit to produce/consume topics starting with `"TOPIC_PREFIX"`:
```
```suggestion
The pulsar add-on given biscuit token can be attenuated, here is an attenuation example using (biscuit-cli)[https://github.com/biscuit-auth/biscuit-cli] from the given biscuit token to produce/consume topics starting with `"TOPIC_PREFIX"`:
```

gaelreyrol replied:
I think you should give an example to "TOPIC_PREFIX" so we don't assume it is an environment variable.

gaelreyrol commented on deploy/addon/pulsar.md:
````diff
@@ -0,0 +1,259 @@
+
+First inspect your biscuit:
+```bash
+$ biscuit inspect addon.biscuit
````
```suggestion
biscuit inspect addon.biscuit
```

gaelreyrol commented on deploy/addon/pulsar.md:
````diff
@@ -0,0 +1,259 @@
+Now attenuate it:
+
+```bash
+$ biscuit attenuate addon.biscuit
````
```suggestion
biscuit attenuate addon.biscuit
```

gaelreyrol commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,259 @@
+
+### Retention
+
+Provided addon pulsar have a infinite retention policies which can be changed using:
```
What are the impacts of an infinite retention policy on Clever Cloud ?
```suggestion
Provided add-on pulsar have an infinite retention policies which can be changed using:
```

gaelreyrol commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,259 @@
+
+Pulsar has a [tiered storage feature](https://pulsar.apache.org/docs/en/tiered-storage-overview/) allowing to offload heavy data to cold storage once a threshold is reached.
+
+For each pulsar addon we provide, we also provide an hidden cellar addon (which is our object storage addon) which is directly binded to the pulsar namespace offload policies. 
```
```suggestion
For each pulsar add-on we provide, we also provide a hidden Cellar add-on (which is our object storage add-on) which is directly binded to the pulsar namespace offload policies. 
```
You should provide a link to the Cellar add-on page.

gaelreyrol commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,259 @@
+* Rust client
+* WebSocket client
+
+### Operations
```
These operations might change in the future ?

KannarFr replied:
yes

gaelreyrol replied:
Could you write something about it ? Our customers can reach us through the support to ask new operations ?

gaelreyrol commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,259 @@
+
+The pulsar addon given biscuit can be attenuated, here is an attenuation example using (biscuit-cli)[https://github.com/biscuit-auth/biscuit-cli] from the given biscuit to produce/consume topics starting with `"TOPIC_PREFIX"`:
+
+First inspect your biscuit:
```
We might need to tell how to create this file ^^.

gaelreyrol reviewed (changes requested):
I think we should stay consistent about Pulsar with a `P` and an add-on with a `-`.
Biscuit is the engine, so we should write `Biscuit` with a `B` and its token: `Biscuit token`.
We should also use the add-on environment variables in all the examples.

Could you revert the format you've made on the `reference-environmemt-variables.md` file, it is unrelated to this PR and unreviewable ^^.

KannarFr commented:
I grouped the suggestions and more related changes in f1317cb9c3032383091a429715fd7670838ee875.

About format, it's already one commit: 2d9225f905febfe313249498a272f9fbacefc99e. I also rebased all cleanly so it can be merge without squash and no problem for the format as it's a commit itself. Is that ok for you?

About the version, I am not sure it's a good to precise it in the documentation as we should manage it each time we bump the version.

KannarFr commented:
About the release pulsar pricing component, @Galimede can you do it if it's ready?

gaelreyrol reviewed (changes requested):
It still misses out a few things but it's much better ! Thanks :+1:

gaelreyrol commented:
> About the version, I am not sure it's a good to precise it in the documentation as we should manage it each time we bump the version.
Could you at least, write some words about it ? What is the release process ? How can they get the version themselves ? I am already anticipating customers reaching the support to ask about the version ^^

PierreZ commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,268 @@
+
+## Overview
+
+Pulsar works on a publisher/subscriber model, just like RabbitMQ or Kafka.
```
Could you add a link to Pulsar's website?
```suggestion
[Pulsar](https://pulsar.apache.org/) works on a publisher/subscriber model, just like RabbitMQ or Kafka.
```

PierreZ commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,268 @@
+## Overview
+
+Pulsar works on a publisher/subscriber model, just like RabbitMQ or Kafka.
+
```
Before talking consuming strategies, we could talk on:

* why are we using pub/sub tech? 
* why do we love event-driven architecture?

KannarFr replied:
@PierreZ Can you contribute these points?

KannarFr replied:
I've done it on 542d67ed3e05f37baf5ca9b17339323f153998b4 and 1ae3ab7a3b8252957823202075500eed6067c30c.

PierreZ commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,268 @@
+
+### Attenuation
+
+The Pulsar add-on given Biscuit token can be attenuated, here is an attenuation example using (biscuit-cli)[https://github.com/biscuit-auth/biscuit-cli] from the given Biscuit token to produce/consume topics starting with `"TOPIC_PREFIX"`:
```
Could you add an attenuation example?

For example, "I want to derive a biscuit from the one I have that can only read data"

Keksoj replied:
I think the attenuation example is already provided.

KannarFr replied:
It is.

PierreZ reviewed (changes requested):
Nice work @Keksoj! I added a few comments.

Keksoj commented:
Okay, I :

- reverted the formatting of the environment variable reference
- changed all generic variable mentions, like <BISCUIT>, to the actual env. variables, like $PULSAR_ADDON_TOKEN
- changed the prefix in the example to a more understandable syntax

Yet to do: 

- [ ] add some context to why we like pub/sub and event driven architecture
- [ ] pricing 
- plus some other things listed accross the PR

gaelreyrol commented:
> Yet to do:
>     * [ ]  pricing

I just created a pull request to update cellar and fsbucket plans: https://github.com/CleverCloud/doc.clever-cloud.com/pull/662.
You can use the same shortcode with `pulsar` like so: `{{< pricingConsumption "pulsar" >}}`.

KannarFr commented:
@Keksoj you are using an old pulsar, the correct env var is `ADDON_PULSAR_TOKEN`, can you fix it? Delete your current addon and create a new one. Sorry about that.

Keksoj commented:
@KannarFr sure !

Keksoj commented:
ping me when you're done, I've got some cosmetic changes to do.

KannarFr commented:
@Keksoj I did it.

@gaelreyrol can you rereview and eventually merge? Then I will bump the related blogpost.

Keksoj commented:
@KannarFr yup!
I will ping @gaelreyrol when done

KannarFr commented:
Oh, the Authorization with Attenuation part must be before the Usage IMHO.

Keksoj commented:
@gaelreyrol Ready for review!

gaelreyrol reviewed (changes requested):
One last thing, you should only use single `\` in your bash examples.

BlackYoup commented:
I'll review this at some point, but don't wait for my review if this is blocking. Just a question: should we start adding CLI references here and there in the documentation? For example, in the `Create add-on` section, we could add `clever addon create pulsar ....`.

I'm starting to write the Jenkins documentation and though about this.

BlackYoup commented:
> I'll review this at some point, but don't wait for my review if this is blocking. Just a question: should we start adding CLI references here and there in the documentation? For example, in the `Create add-on` section, we could add `clever addon create pulsar ....`.
> 
> I'm starting to write the Jenkins documentation and though about this.

I just saw that there is a partial `create-application.md` that display a CLI command. Maybe we could have the same type of partial for add-ons. What do you think?

gaelreyrol commented:
> > I'll review this at some point, but don't wait for my review if this is blocking. Just a question: should we start adding CLI references here and there in the documentation? For example, in the `Create add-on` section, we could add `clever addon create pulsar ....`.
> > I'm starting to write the Jenkins documentation and though about this.
> 
> I just saw that there is a partial `create-application.md` that display a CLI command. Maybe we could have the same type of partial for add-ons. What do you think?

That's a great idea ! :slightly_smiling_face:

PierreZ commented:
Nice work :+1: :rocket:

BlackYoup commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,413 @@
+* Refreshing distributed caches. For example, an application can publish invalidation events to update the IDs of objects that have changed.
+* Real-time event distribution Events, raw or processed, may be made available to multiple applications across your team and organization for real time processing.
+
+## Create an add-on
```
```suggestion
## Create a Pulsar add-on
```

BlackYoup commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,413 @@
+
+## Common use cases
+
+* Ingestion user interaction and server events To make use of user interaction events from end-user apps or server events from your system, you may forward them to Pulsar and then use a stream processing tool which delivers them to your applications. Pulsar allows you to gather events from many clients simultaneously.
```
I can't really understand this sentence, am I the only one?

BlackYoup commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,413 @@
+
+Tenants and namespaces allow for grouping and subgrouping of topics.
+
+A Clever Cloud Pulsar add-on is basically a immutable `tenant/namespace` where the tenant is your user id, and the namespace is the add-on id.
```
`an immutable`?

BlackYoup commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,413 @@
+
+* Ingestion user interaction and server events To make use of user interaction events from end-user apps or server events from your system, you may forward them to Pulsar and then use a stream processing tool which delivers them to your applications. Pulsar allows you to gather events from many clients simultaneously.
+* Replicating data among databases using [Pulsar IO](https://pulsar.apache.org/docs/en/io-overview/) is commonly used to distribute change events from databases.
+* Parallel processing and workflows. You can efficiently distribute a large number of tasks among multiple workers ( compressing text files, sending email notifications).
```
```suggestion
* Parallel processing and workflows. You can efficiently distribute a large number of tasks among multiple workers (compressing text files, sending email notifications).
```

BlackYoup commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,413 @@
+
+We maintain up-to-date Pulsar clusters based on the official Apache Pulsar release process. Your Pulsar add-on version is available in your add-on dashboard. Updates will be notified using Clever Cloud weekly updates on the [blog](https://www.clever-cloud.com/blog/).
+
+## Common use cases
```
Almost of the items of the list use a different "example" pattern. Maybe make all the elements like:
`- <example>: explain it`

KannarFr replied:
Well no, it is too complicated to make generic as some are parts of other etc.

KannarFr commented:
> One last thing, you should only use single `\` in your bash examples.

fixed in 17280cd860b2085f150946451291262a2d9310ad.

@BlackYoup I agree with you, as we need to release the blogpost fast, can you file an issue?

@gaelreyrol can you merge, release and publish this?

BlackYoup commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,413 @@
+
+### Attenuation
+
+The Pulsar add-on given Biscuit token can be attenuated, here is an attenuation example using [biscuit-cli](https://github.com/biscuit-auth/biscuit-cli) from the given Biscuit token to produce/consume topics starting with a custom topic prefix called `"my-own-prefix"`.
```
```suggestion
The Biscuit provided by a pulsar add-on can be attenuated. Here is an attenuation example using [biscuit-cli](https://github.com/biscuit-auth/biscuit-cli) to produce/consume topics starting with a custom topic prefix called `"my-own-prefix"`.
```

Also, should Biscuit put between back ticks? `Biscuit`

BlackYoup commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,413 @@
+
+The Pulsar add-on given Biscuit token can be attenuated, here is an attenuation example using [biscuit-cli](https://github.com/biscuit-auth/biscuit-cli) from the given Biscuit token to produce/consume topics starting with a custom topic prefix called `"my-own-prefix"`.
+
+Put your Biscuit token in a file:
```
Write your biscuit token?

BlackYoup commented on deploy/addon/pulsar.md:
````diff
@@ -0,0 +1,413 @@
+==========
+```
+
+Now the block n°2 ensure that topics must starts with `"my-own-prefix"`.
````
```suggestion
Now the block n°2 ensures that topics must start with `"my-own-prefix"`.
```

BlackYoup commented on deploy/addon/pulsar.md:
````diff
@@ -0,0 +1,413 @@
+          topics list $ADDON_PULSAR_TENANT/$ADDON_PULSAR_NAMESPACE
+```
+
+As Biscuit is a token, you can use `AuthenticationToken($ADDON_PULSAR_TOKEN)` provided by [clients libraries](https://pulsar.apache.org/docs/en/client-libraries/) to authenticate to our clusters without any tweak.
````
I think this should go at the end of the `## Authorization` section, before the `### Attenuation` one

BlackYoup commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,413 @@
+
+### Operations
+
+The pulsar addon given Biscuit token enables you to run several operations on namespace, its policies and the related topics.
```
```suggestion
The Biscuit token provided by the Pulsar add-on allows you to run several operations on namespace, its policies and the related topics.
```

BlackYoup commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,413 @@
+
+The pulsar addon given Biscuit token enables you to run several operations on namespace, its policies and the related topics.
+
+These operations might change in the future. Don't hesitate to call our support to ask for new operations!
```
```suggestion
These operations might change in the future. Don't hesitate to write to our support to ask for new operations!
```

BlackYoup commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,413 @@
+
+### Offload storage to Cellar (S3)
+
+Pulsar has a [tiered storage feature](https://pulsar.apache.org/docs/en/tiered-storage-overview/) allowing to offload heavy data to cold storage once a threshold is reached.
```
```suggestion
Pulsar has a [tiered storage feature](https://pulsar.apache.org/docs/en/tiered-storage-overview/) allowing offload of heavy data to cold storage once a threshold is reached.
```

BlackYoup commented on deploy/addon/pulsar.md:
```diff
@@ -0,0 +1,413 @@
+
+Pulsar has a [tiered storage feature](https://pulsar.apache.org/docs/en/tiered-storage-overview/) allowing to offload heavy data to cold storage once a threshold is reached.
+
+For each Pulsar add-on we provide, we also provide an hidden [Cellar add-on](https://www.clever-cloud.com/doc/deploy/addon/cellar/) (which is our object storage addon) which is directly binded to the Pulsar namespace offload policies.
```
```suggestion
For each Pulsar add-on we provide, we also provide a hidden [Cellar add-on]({{< ref "deploy/addon/cellar.md" >}}), our object storage add-on. This Cellar add-on is bound to the Pulsar namespace offload policies and will store your offloaded data.
```

BlackYoup reviewed (changes requested):
Let me know what you think of those requests :)

KannarFr commented:
@BlackYoup I applied almost all of your suggestions. @gaelreyrol gogo! :p

gaelreyrol commented:
Nice work guys ! :clap:
